precise/iaf_psc_exp_ps_time_reversal.h

```diff
@@ -0,0 +1,589 @@
+  stmp.set(d, ptmp, delta_EL);      // throws if BadProperty
+
+  // if we get here, temporaries contain consistent set of properties
+  P_ = ptmp;
+  S_ = stmp;
+}
+
+
+// inline
+// void iaf_psc_exp_ps_time_reversal::print_stopwatch()
+// {
+//   regime1.print("branch-1 (no-spike) took: ");
+//   regime2.print("branch-2 (no-spike) took: ");
+//   regime3.print("branch-3 (spike) took: ");
+
+// }
+
+inline
+void nest::iaf_psc_exp_ps_time_reversal::spike_test_(const double_t t1)
+{
+  // we assume that P_.tau_ex_=P_.tau_in_
+  double_t const I_0   = V_.y1_ex_before_ + V_.y1_in_before_;
+  double_t const V_0   = V_.y2_before_;
+  double_t const I_t1  = S_.y1_ex_ + S_.y1_in_;
+  double_t const V_t1  = S_.y2_;
+  double_t const tau   = P_.tau_ex_;
+  double_t const tau_m = P_.tau_m_;
+  double_t const I_x   = P_.I_e_;
+  double_t const C_m   = P_.c_m_;
+  double_t const V_th  = P_.U_th_;
+
+  double_t const tauC_m = tau_m/C_m;
+
+  double_t const Vdot_t1 = -V_t1/tau_m + (I_t1+I_x)/C_m;
+
+
+  // iaflossless tests
+  if ( Vdot_t1 < 0.0 )
+  {
+    double_t const Vdot_0  = -V_0/tau_m + (I_0+I_x)/C_m;
+
+    if ( Vdot_0 > 0.0 )
+    {
+      if ( Vdot_0*t1 + V_0 >= V_th )
+      {
+
+  	  // D'Haene tests
+  	  double_t const minus_taus = -tau_m*tau / (tau_m-tau);
+  	  double_t const V_syn      = minus_taus / C_m * I_0;
+  	  double_t const V_m        = V_0 - tauC_m*I_x - V_syn;
+  	  double_t const quot       = -tau*V_m / (tau_m*V_syn);
+
+  	  double_t const t_max = minus_taus * log(quot);
+
+  	  double_t const expm1_tau_syn = numerics::expm1(-t_max/tau);
+  	  double_t const expm1_tau_m   = numerics::expm1(-t_max/tau_m);
+      
+  	  double_t const P20 = -tau_m*expm1_tau_m / C_m;
+  	  double_t const P21 = minus_taus / C_m * (expm1_tau_syn-expm1_tau_m);
+      
+  	  if ( (P20*I_x + P21*I_0 + expm1_tau_m*V_0 + V_0) >= V_th )
+  	    S_.dhaene_det_spikes++;
+
+	  //}
+      }
+    }
+  }
+}
+
+
+/* Conventional spike detection algorithms propagate the initial state forwards in time and see whether it meets the threshold.
+This function implements a general method to solve the threshold-detection problem for an integrable, affine or linear
+time evolution by applying geometric analysis. The idea is to propagate the threshold backwards in time and see whether
+it meets the initial state. In state space spanned by voltage and current, this clearly separates the spiking region and 
+non-spiking region. is_spike_ takes argument dt which corresponds to the time window at which this spike prediction occurs.
+returns true, spike: if (V(t_{right}) > V_(\theta)); returns false: (V(t_{right} < V_(\theta) or initial conditions in no-spike region;
+returns true, spike: missed spike excursion, compute t_{max} = dt and find point of threshold crossing t_{\theta} using emit_spike_.
+inequalities are adjusted such that backward propagation (negative time) is already accounted for here */
+
+inline
+bool iaf_psc_exp_ps_time_reversal::is_spike_(double_t dt)
```

Comment:
This method can be `const`, I believe. Move it to the `cpp` file, since it is only called from the `cpp` file. It is also so complex that there is no point in inlining it.